File: 98-validate-binary-search-tree/98-validate-binary-search-tree.java
class Solution {
        public boolean isValidBST(TreeNode node) {
        return isValid(node, null, null);
    }
    
    private boolean isValid(TreeNode n, Integer min, Integer max) {
        return  (n == null) ||
                (min == null || n.val > min) && (max == null || n.val < max) &&
                isValid(n.left , min , n.val) && isValid(n.right, n.val, max);
    }
}

File: 98-validate-binary-search-tree/98-validate-binary-search-tree.swift
class Solution {
    func isValidBST(_ root: TreeNode?) -> Bool {
        return isBst(root, min: Int.min, max: Int.max)
    }
    
    private func isBst(_ node: TreeNode?, min: Int, max: Int) -> Bool {
        if node == nil { return true }
        if node!.val <= min || node!.val >= max { return false }
        return isBst(node?.left, min: min, max: node!.val) && isBst(node?.right, min: node!.val, max: max)
    }
}